src/Writer.php

```diff
@@ -0,0 +1,215 @@
+    }
+
+    /**
+     * Write a wiki page into the PDF
+     *
+     * @param string $html The rendered HTML of the wiki page
+     * @return void
+     * @throws MpdfException
+     */
+    public function wikiPage(string $html): void
+    {
+        $this->conditionalPageBreak();
+
+        $this->applyHeaderFooters();
+
+        $this->mpdf->WriteHTML($html, HTMLParserMode::HTML_BODY, false, false);
+
+        // add citation box if any
+        $cite = $this->template->getHTML('citation');
+        if ($cite) {
+            $this->mpdf->WriteHTML($cite, HTMLParserMode::HTML_BODY, false, false);
+        }
+
+        $this->breakAfterMe();
+    }
+
+    public function toc(): void
+    {
+        $this->conditionalPageBreak();
+        // FIXME
+        $this->breakAfterMe();
+    }
+
+    /**
+     * Insert a cover page
+     *
+     * Should be called once at the beginning of the PDF generation. Will do nothing if
+     * no cover page is configured.
+     *
+     * @return void
+     * @throws MpdfException
+     */
+    public function cover(): void
+    {
+        $this->conditionalPageBreak();
+
+        $html = $this->template->getHTML('cover');
+        if (!$html) return;
+
+        $this->mpdf->WriteHTML($html, HTMLParserMode::HTML_BODY, false, false);
+
+        $this->breakAfterMe();
+    }
+
+    /**
+     * Insert a back page
+     *
+     * Should be called once at the end of the PDF generation. Will do nothing if
+     * no back page is configured.
+     *
+     * @return void
+     * @throws MpdfException
+     */
+    public function back(): void
+    {
+        $this->conditionalPageBreak();
+
+        $html = $this->template->getHTML('back');
+        if (!$html) return;
+
+        $this->mpdf->WriteHTML($html, HTMLParserMode::HTML_BODY, false, false);
+    }
+
+    /**
+     * Finalize the document
+     *
+     * @return void
+     * @throws MpdfException
+     */
+    public function endDocument(): void
+    {
+        // adds the closing div and finalizes the document
+        $this->mpdf->WriteHTML('</div>', HTMLParserMode::HTML_BODY, false, true);
+    }
+
+    /**
+     * Set new headers and footers
+     *
+     * This will call the appropriate mpdf methods to set headers and footers. It should be called
+     * before each wiki page is added to the PDF.
+     *
+     * On first call on this instance it will set the headers/footers for the first page, afterwards
+     * it will use the standard headers/footers.
+     *
+     * We always set even and odd headers/footers, though they may be identical.
+     * @return void
+     */
+    protected function applyHeaderFooters(): void
+    {
+        if ($this->isFirstPage) {
+            $header = $this->template->getHTML('header', 'first');
+            $footer = $this->template->getHTML('footer', 'first');
+
+            if ($header) {
+                $this->mpdf->SetHTMLHeader($header, 'O');
+                $this->mpdf->SetHTMLHeader($header, 'E');
```

Comment:
I did not expect both O and E need to be set for a first page. Does this mean also a second page in the document is always affected? (not tested yet)
(the old code used named html with pseudo css, difficult to see quickly what happened before)

Reply:
good question. I guess the first page is always odd? OTOH the first page might be pushed to an even page if there's a TOC before? I think it's okay to set both - mpdf will pick what it needs.